OpenRA.Game/WorldUtils.cs

```diff
@@ -41,6 +41,113 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			{
+				long actorWidth;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+					actorWidth = healthInfo.Radius.Length;
+				else
+					actorWidth = 0;
+
+				var distance = MinimumDistancePointAndVector(target, source, currActor.CenterPosition);
+				var maxReach = (long)(actorWidth + rayWidth.Length);
+				maxReach *= maxReach;
+
+				// Log.Write("debug", "FindActorsInRaycast, distance : {0}, maxReach : {1}", distance, maxReach);
+				if (distance <= maxReach)
+					intersectedActors.Add(currActor);
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Returns the (squared) minimum distance between a line and a point.
+		/// From http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+		/// </summary>
+		/// <param name="target">The target point (head) of the line</param>
+		/// <param name="source">The source point (tail) of the line</param>
+		/// <param name="point">The point that the minimum distance should be found to</param>
+		/// <returns>The square of the minimum distance between the line and point.</returns>
+		private static long MinimumDistancePointAndVector(WPos target, WPos source, WPos point)
+		{
+			// Return minimum distance between line segment target->source and point
+			var squaredLength = (target - source).LengthSquared;
+
+			// target = source case, so just find distance from the point to the target.
+			if (squaredLength == 0.0) return (target - point).LengthSquared;
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			long t = DotProduct(target, source, point);
+
+			// Beyond the 'target' end of the segment
+			if (t < 0)
+				return (point - target).LengthSquared;
+
+			// Beyond the 'source end of the segment
+			if (t > squaredLength)
+				return (point - source).LengthSquared;
+
+			// Projection falls on the segment
+			// Using floating points, this would be: var newX = (int)((double)target.X + t * (double)(source.X - target.X));
+			// where t would already be divided by squaredLength.
+			var newX = (int)(target.X + (t * (source.X - target.X) / squaredLength));
+			var newY = (int)(target.Y + (t * (source.Y - target.Y) / squaredLength));
+			var projection = new WPos(newX, newY, (target.Z + source.Z) / 2);
+			return (point - projection).LengthSquared;
+		}
+
+		private static long DotProduct(WPos target, WPos source, WPos point)
+		{
+			var xDiff = (point.X - target.X) * (source.X - target.X);
+			var yDiff = (point.Y - target.Y) * (source.Y - target.Y);
+
+			// Should be dist = (target - source).LengthSquared;
+			// But in order to avoid floating points, we do not divide here, but rather work with the large numbers as far as possible.
+			var dist = 1;
+			var ans = (xDiff + yDiff) / dist;
```

Comment:
Don't you want `xDiff + yDiff`? If all you are trying to do is make the calculation return a `long`, then `xDiff + yDiff` will do the job since it is an `int` and `int`s are implicitly converted to `long`s. In fact `return xDiff + yDiff` can replace `var dist = 1;`...`return ans;`

What you might actually want is:
```c#
var xDiff = ((long)point.X - target.X) * (source.X - target.X);
var yDiff = ((long)point.Y - target.Y) * (source.Y - target.Y);
return xDiff + yDiff;
```
- OR -
```c#
var xDiff = ((long)point.X - target.X) * ((long)source.X - target.X);
var yDiff = ((long)point.Y - target.Y) * ((long)source.Y - target.Y);
return xDiff + yDiff;
```
- OR -
```c#
var xDiff = ((long)point.X - (long)target.X) * ((long)source.X - (long)target.X);
var yDiff = ((long)point.Y - (long)target.Y) * ((long)source.Y - (long)target.Y);
return xDiff + yDiff;
```
since `long` is used as the value type of `LengthSquared` to avoid multiplication overflow.
The first one works and is shortest, the second one is like the body of `WVec.LengthSquared` get, and the last one makes the fact that all of the arithmetic is with `long`s.

Are you assuming that these values are `float`s or `double`s?
`WVec` and `WPos` use `int`s to store world units (1024/cell).